src/ray/core_worker/core_worker_process.cc

```diff
@@ -423,8 +423,23 @@ std::shared_ptr<CoreWorker> CoreWorkerProcessImpl::CreateCoreWorker(
       /*put_in_local_plasma_callback=*/
       [this](const RayObject &object, const ObjectID &object_id) {
         auto core_worker = GetCoreWorker();
-        RAY_CHECK_OK(
-            core_worker->PutInLocalPlasmaStore(object, object_id, /*pin_object=*/true));
+
+        auto status =
+            core_worker->PutInLocalPlasmaStore(object, object_id, /*pin_object=*/true);
+        if (!status.ok()) {
+          // During shutdown, plasma store connections can be closed causing IOError.
+          // Tolerate only IOError during shutdown to avoid masking real errors.
+          if (status.IsIOError() && core_worker->IsShuttingDown()) {
+            // Double-check shutdown state - this handles the race where shutdown
+            // began after our first check but before the plasma operation.
+            RAY_LOG(WARNING) << "Failed to put error object " << object_id
+                             << " in plasma store during shutdown: " << status.ToString();
```

Comment:
This still feels a bit dangerous to me because the callsites will assume that the plasma operation succeeded, when it really silently failed. This might cause check failures in other places in the stack or other undefined behavior due to a broken invariant.

It's unclear to me that it is a significant improvement over what we have now, which is equally incorrect but at least a little bit more explicit.

I'm OK with it as a temporary solution, but only if we intend to follow up and propagate the status appropriately and handle it at the callsites.

Reply:
I already tightened the current behavior to only tolerate IOError during shutdown and made the log explicit that the object was not stored. That said, I totally agree with the concern. Maybe we can add a counter/metric `plasma_put_failed_due_to_shutdown` so we can alert/observe this path.

To fully address this, I'll follow up by changing the `PutInLocalPlasmaCallback` to return Status and propagate that through `TaskManager`. For failures that are not (shutdown + IOError), `TaskManager` will mark the task failed with an unexpected/system error. This ensures callsites don't assume success on a failed plasma put. I'll also audit other Raylet IPC/plasma callsites for the same pattern.